### ISSUE TITLE
Re: Cubature goes outside the bounds of integration box

### ISSUE THREAD

**Issue by ohannuks** [saullocastro/cubature#17]
Hi! Sorry to bother you; I am not sure if you still work on this, but I noticed that cubature seems to try to take values which are outside the integration domain at times. In particular, I made a simple integration script setting (xmin,xmax) = (125.2414,243124) which raises an error if the integrand tries to evaluate values which are outside the integration domain. In this specific example, the integrator tries to insert "1" into the integrand, which is outside the bound of the integration domain. 

See the attached script:

```
from cubature import cubature
import numpy as np

xmin =np.array([125.24214])
xmax =np.array([243124.0])

def testintegrand_v(x):
  x = np.array(x)
  return x[:,0]**2

numerical  = cubature(testintegrand_v,1,1,xmin,xmax,vectorized=True, abserr=1.e-8,relerr=1.e-8)[0]
analytical = (xmax**3-xmin**3)/3.
print "Analytical vs numerical (looks ok): %f %f"%(numerical,analytical)

# Same but now testing also for xmin and xmax bounds

def testintegrand_v(x):
  x = np.array(x)
  if np.min(x) < xmin:
    print x
    raise ValueError("x is out of bounds with x = %f and xmin = %f; see the above print-out for x"%(x[np.argmin(x)],xmin))
  return x[:,0]**2
cubature(testintegrand_v,1,1,xmin,xmax,vectorized=True, abserr=1.e-8,relerr=1.e-8)[0]
```

The output is:
```

otto@otto:/tmp$ python cubatest.py 
Analytical vs numerical (looks ok): 4790294812348709.000000 4790294812348709.000000
[[ 1.]
 [ 1.]
 [ 1.]
 [ 1.]
 [ 1.]
 [ 1.]
 [ 1.]]
Traceback (most recent call last):
  File "cubatest.py", line 23, in <module>
    cubature(testintegrand_v,1,1,xmin,xmax,vectorized=True, abserr=1.e-8,relerr=1.e-8)[0]
  File "build/bdist.linux-x86_64/egg/cubature/cubature.py", line 227, in cubature
  File "cubatest.py", line 21, in testintegrand_v
    raise ValueError("x is out of bounds with x = %f and xmin = %f; see the above print-out for x"%(x[np.argmin(x)],xmin))
ValueError: x is out of bounds with x = 1.000000 and xmin = 125.242140; see the above print-out for x
```

I currently have cubature installed via the master branch @ commit:

```
commit 6dfca1349dd585ff9e6ced7d244be5c124d801eb
Merge: f6595a4 f87d3fe
Author: saullocastro <saullogiovani@gmail.com>
Date:   Fri Feb 19 03:33:00 2016 -0200

    Merge pull request #16 from markdewing/callback
    
    Add ctypes-based callbacks.
```

Edit:
The reason I noticed this is the first time was because one of my integrand included a square root of a certain expression which would be negative outside the integration bounds and hence the result was nan when cubature tried to evaluate one point at x=np.ones(ndim).

**Comment by saullocastro** [comment]
Dear Otto

I understood the error description. I'll have a look at it to see what can
be done.

Regards
Saullo

On May 28, 2017 02:51, "Otto Akseli Hannuksela" <notifications@github.com>
wrote:

Hi! Sorry to bother you; I am not sure if you still work on this, but I
noticed that cubature seems to try to take values which are outside the
integration domain at times. In particular, I made a simple integration
script setting (xmin,xmax) = (125.2414,243124) which raises an error if the
integrand tries to evaluate values which are outside the integration
domain. In this specific example, the integrator tries to insert "1" into
the integrand, which is outside the bound of the integration domain.

See the attached script:

from cubature import cubature
import numpy as np

xmin =np.array([125.24214])
xmax =np.array([243124.0])

def testintegrand_v(x):
  x = np.array(x)
  return x[:,0]**2

numerical  = cubature(testintegrand_v,1,1,xmin,xmax,vectorized=True,
abserr=1.e-8,relerr=1.e-8)[0]
analytical = (xmax**3-xmin**3)/3.
print "Analytical vs numerical (looks ok): %f %f"%(numerical,analytical)

# Same but now testing also for xmin and xmax bounds

def testintegrand_v(x):
  x = np.array(x)
  if np.min(x) < xmin:
    print x
    raise ValueError("x is out of bounds with x = %f and xmin = %f;
see the above print-out for x"%(x[np.argmin(x)],xmin))
  return x[:,0]**2
cubature(testintegrand_v,1,1,xmin,xmax,vectorized=True,
abserr=1.e-8,relerr=1.e-8)[0]

The output is:


otto@otto:/tmp$ python cubatest.py
Analytical vs numerical (looks ok): 4790294812348709.000000
4790294812348709.000000
[[ 1.]
 [ 1.]
 [ 1.]
 [ 1.]
 [ 1.]
 [ 1.]
 [ 1.]]
Traceback (most recent call last):
  File "cubatest.py", line 23, in <module>
    cubature(testintegrand_v,1,1,xmin,xmax,vectorized=True,
abserr=1.e-8,relerr=1.e-8)[0]
  File "build/bdist.linux-x86_64/egg/cubature/cubature.py", line 227,
in cubature
  File "cubatest.py", line 21, in testintegrand_v
    raise ValueError("x is out of bounds with x = %f and xmin = %f;
see the above print-out for x"%(x[np.argmin(x)],xmin))
ValueError: x is out of bounds with x = 1.000000 and xmin =
125.242140; see the above print-out for x

I currently have cubature installed via the master branch @ commit:

commit 6dfca1349dd585ff9e6ced7d244be5c124d801eb
Merge: f6595a4 f87d3fe
Author: saullocastro <saullogiovani@gmail.com>
Date:   Fri Feb 19 03:33:00 2016 -0200

    Merge pull request #16 from markdewing/callback

    Add ctypes-based callbacks.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/saullocastro/cubature/issues/17>, or mute the thread

**Comment by antonl** [comment]
@saullocastro I think this may be due to the testing of the integrand function we do around
https://github.com/saullocastro/cubature/blob/master/cubature/cubature.py#L227
The value of the integrand is not actually used for anything, just to determine the size of the output vector. We could perhaps map the tested values to the (min,max) domain to avoid this issue.

@ohannuks Is the calculated integral value incorrect for the sqrt integrand?

**Comment by saullocastro** [comment]
@ohannuks  I agree with @antonl. I did a fix and now release 0.13.3 should work just fine.

**Comment by ohannuks** [comment]
@antonl  The integral value is correct; I was just puzzled since I previously had error checking inside the integrand to make sure I was not getting nans.


@saullocastro  Many thanks! That's perfect!
